JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java

```diff
@@ -10,91 +10,329 @@
+		formData.put("username",      prop.getProperty("User"));
+		formData.put("password",      prop.getProperty("Password"));
+        GetAndSetTokens(formData);
     }
-
-    private static void printResults(ResultSet rs) throws SQLException {
-        int rowIdx = 1;
-        while (rs.next()) {
-            System.out.println(rowIdx + ". " + rs.getString(1).trim() + " " + rs.getString(2).trim());
-            rowIdx++;
+    // refresh grant uses the refresh token to get a new access and refresh token
+    public static void GetTokensByRefreshGrant() throws Exception {
+    	Map<String, String> formData = new HashMap<String, String>();
+   		formData.put("grant_type",		"refresh_token");
+		formData.put("client_id",		prop.getProperty("ClientId"));
+		formData.put("client_secret",	prop.getProperty("ClientSecret"));
+		formData.put("refresh_token",	System.getProperty(OAUTH_REFRESH_TOKEN_VAR_STRING));
+		GetAndSetTokens(formData);
+    }
+    // read result from Buffered input stream
+	private static ByteArrayOutputStream readResult(BufferedInputStream in, ByteArrayOutputStream buf) {
+		try
+		{
+			for (int result = in.read(); result != -1; result = in.read()) 
+			{
+				buf.write((byte) result);
+			}
+		}catch(Exception e)
+		{
+			e.printStackTrace();
+		}
+		return buf;
+	}
+    // get and set tokens from IDP 
+	private static void GetAndSetTokens(Map<String, String> formData) throws Exception
+    {
+    	try {
+	    	String postOpts = getEncodedParamsString(formData);
+	        byte[] postData = postOpts.getBytes("UTF-8");
+	        int postDataLength = postData.length;
+	        URL url = new URL(prop.getProperty("TokenUrl"));
+	        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+	        try {
+	            connection.setDoOutput(true);
+	            connection.setUseCaches(false);
+	            connection.setRequestMethod("POST");
+	            connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
+	            connection.setRequestProperty("Content-Length", Integer.toString(postDataLength));
+	            connection.setRequestProperty("Accept", "application/json");
+				connection.getOutputStream().write(postData);
+	           	BufferedInputStream in = new BufferedInputStream(connection.getInputStream());
+	    	    ByteArrayOutputStream buf = new ByteArrayOutputStream();
+	            readResult(in, buf);
+	    	    String res = buf.toString("UTF-8");
+	        	JsonElement jElement = JsonParser.parseString(res);
+	            JsonObject jObject = jElement.getAsJsonObject();
+				// set Tokens as System Properties - new values to access_token and refresh_token
+	            String accessToken = jObject.has("access_token") ? jObject.get("access_token").getAsString() : null;
+	            String refreshToken = jObject.has("refresh_token") ? jObject.get("refresh_token").getAsString() : null;           
+	            if (null == accessToken || null == refreshToken)
```

Comment:
It is possible that an IDP will return an access token but not a refresh token. If the refresh token is null, that could be OK, it just shouldn't be set in the environment variable.

Reply:
Yes, It's possible. As in our application, We are showing the use of access and refresh tokens, So let's have both tokens. This sample app shows how we can use the access and refresh token. If any User wants to have code without a refresh token, They can remove the code and use it as it's open-source code. Adding multiple scenarios like this will increase the complexity of the sample app.